### PR TITLE
zephyr: Kconfig: SOC_GECKO_USE_RAIL_API option for RAIL for proprietary

### PR DESCRIPTION
**Pull request by zohavas**
Currently on zephyr, RAIL library is used only by Bluetooth applications,
with this update, it will be able to be used for sample applications 
for custom radio phys.
All files were copied from Silicon Labs GSDK v4.2.4.

### CODE REVIEW

**Comment by zohavas** [comment]
@fkokosinski This is the other part of our PR for Zephyr.

File: gecko/platform/radio/rail_lib/plugin/rail_util_init/config/proprietary/sl_rail_util_init_inst0_config.h [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+									       *such, and must not
+									       *be misrepresented as
+									       *being the original
+									       *software.
+									       * 3. This notice may
+									       *not be removed or
+									       *altered from any
+									       *source distribution.
+									       *
+									       ******************************************************************************/
```
Comment (jerome-pouiller) [inline gecko/platform/radio/rail_lib/plugin/rail_util_init/config/proprietary/sl_rail_util_init_inst0_config.h:71]: hu?

File: gecko/CMakeLists.txt [hunk 2]
```diff
@@ -37,7 +37,7 @@ endif()
 
 if (${CONFIG_BOARD} STREQUAL "efr32bg27_brd2602a")
   set(SILABS_GECKO_BOARD "brd2602a")
-  set(PRODUCT_NO "sdid205")
+  set(PRODUCT_NO "sdid230")
```
[the remark this answers is not in the document]
Comment (zohavas) [reply gecko/CMakeLists.txt:40]: Thank, you I will update the PR accordingly.

File: gecko/CMakeLists.txt [hunk 3]
```diff
@@ -49,6 +49,7 @@ zephyr_include_directories(
   platform/radio/rail_lib/chip/efr32/efr32xg2x
   platform/radio/rail_lib/common
   platform/radio/rail_lib/plugin/pa-conversions
+  platform/radio/rail_lib/plugin/pa-conversions/efr32xg${GECKO_SERIES_NUMBER}
   platform/radio/rail_lib/plugin/pa-conversions/efr32xg${GECKO_SERIES_NUMBER}/config
```
[the remark this answers is not in the document]
Comment (zohavas) [reply gecko/CMakeLists.txt:53]: When I created it, there were no stable support for xg27, and after talking to apps team, and looking into Zephyr support we agreed that we won't support series 1 boards.

File: gecko/CMakeLists.txt [hunk 4]
```diff
@@ -162,3 +163,24 @@ if(CONFIG_SOC_GECKO_USE_RAIL)
...
+    platform/radio/rail_lib/plugin/rail_util_protocol
+    platform/radio/rail_lib/plugin/rail_util_protocol/config/efr32xg${GECKO_SERIES_NUMBER}/
+    platform/radio/rail_lib/protocol/ble
+    platform/radio/rail_lib/protocol/ieee802154
+    platform/radio/rail_lib/protocol/zwave
+  )
+  
+  zephyr_sources(platform/radio/rail_lib/plugin/rail_util_protocol/sl_rail_util_protocol.c)
+  zephyr_sources(platform/radio/rail_lib/plugin/rail_util_init/autogen/proprietary/sl_rail_util_init.c)
+  zephyr_sources(platform/radio/rail_lib/plugin/rail_util_callbacks/autogen/proprietary/sl_rail_util_callbacks.c)
```
[the remark this answers is not in the document]
Comment (zohavas) [reply gecko/CMakeLists.txt:184]: No these files come from the GSDK, if they are in the autogen folder, that means those are generated from templates file in GSDK. I strongly suggest to keep these files as is, as they can be copied from GSDK or generated GSDK project, and modifying them would mean huge overhead and incompatibility with further updates.

File: gecko/platform/radio/rail_lib/plugin/rail_util_init/config/proprietary/sl_rail_util_init_inst0_config.h (hunk 1 above)
Comment (zohavas) [reply gecko/platform/radio/rail_lib/plugin/rail_util_init/config/proprietary/sl_rail_util_init_inst0_config.h:71]: Accidently the Zephyr code beautifier was run on this file. I will update it.

File: gecko/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (zohavas) [reply gecko/CMakeLists.txt:184]: I will update these files as there were a blob update for rail lib and I would like to keep the versions up to date.
I will update the commit message as well.

File: gecko/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerome-pouiller) [reply gecko/CMakeLists.txt:53]: Do you think there is a way to show these requirements in the Kconfig file? So the user won't get a compilation error if they select `CONFIG_SOC_GECKO_USE_RAIL_API` + `xg27`?

File: gecko/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (zohavas) [reply gecko/CMakeLists.txt:40]: https://github.com/zephyrproject-rtos/hal_silabs/pull/44

File: gecko/platform/radio/rail_lib/plugin/pa-conversions/efr32xg24/config/module/sl_rail_util_pa_config.h
Comment (jerome-pouiller) [inline gecko/platform/radio/rail_lib/plugin/pa-conversions/efr32xg24/config/module/sl_rail_util_pa_config.h:1]: All the other files are located in `.../efr32xg*/config/`. Is this location expected?

Comment (zohavas) [reply gecko/platform/radio/rail_lib/plugin/pa-conversions/efr32xg24/config/module/sl_rail_util_pa_config.h:1]: I am following the original Silabs structure: https://github.com/SiliconLabs/gecko_sdk/blob/gsdk_4.2/platform/radio/rail_lib/plugin/pa-conversions/efr32xg24/config/module/sl_rail_util_pa_config.h

File: gecko/CMakeLists.txt [hunk 5]
```diff
@@ -51,11 +51,13 @@ endfunction()
 if(${CONFIG_SOC_GECKO_HAS_RADIO})
   if(${CONFIG_SOC_GECKO_SERIES1})
```
[the remark this answers is not in the document]
Comment (zohavas) [reply gecko/CMakeLists.txt:52]: Hello @jerome-pouiller,
I understand your point, but I think that these aren't big changes, but if you and @fkokosinski request me to make these into 3 PRs I will happily do it.

File: gecko/CMakeLists.txt [hunk 6]
```diff
@@ -51,11 +51,13 @@ endfunction()
...
     zephyr_include_directories(
+      platform/radio/rail_lib/plugin/pa-conversions/efr32xg1x
       platform/radio/rail_lib/plugin/pa-conversions/efr32xg1x/config
       platform/radio/rail_lib/chip/efr32/efr32xg1x
     )
   elseif(${CONFIG_SOC_GECKO_SERIES2})
     zephyr_include_directories(
+      platform/radio/rail_lib/plugin/pa-conversions/efr32xg${GECKO_SERIES_NUMBER}
       platform/radio/rail_lib/plugin/pa-conversions/efr32xg${GECKO_SERIES_NUMBER}/config
```
Comment (jerome-pouiller) [inline gecko/CMakeLists.txt:61]: It seems the identification of the proper configuration is done by `sl_rail_util_pa_conversions_efr32_config.h` and `pa_curves_efr32.h`. So the two lines above seems useless.

File: gecko/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (fkokosinski) [reply gecko/CMakeLists.txt:52]: Yup - splitting it into 3 commits is a good idea.

**Review by jerome-pouiller — APPROVED** [review]
I assume this state is acceptable for now.

**Review by fkokosinski — APPROVED** [review]
LGTM. Let's wait for the Zephyr PR to have 2 ACKs before merging this one.

**Comment by zohavas** [comment]
@fkokosinski is @jerome-pouiller approve okay for 2 ACKs or should I start to ping other code owners as well?

**Comment by fkokosinski** [comment]
> @fkokosinski is @jerome-pouiller approve okay for 2 ACKs or should I start to ping other code owners as well?

Hi! The 2 ACK is not a strict Zephyr requirement. It's more for me (or any other maintainer, really), so that we know that the changes in the main Zephyr repo won't be reviewed negatively, or even rejected. I suppose this one has been sitting long enough that we know that there isn't any opposition to it in Zephyr, really.

**Pull request merged**